### PR TITLE
[R-package] introduce Dataset methods set_field() and get_field()

### PR DESCRIPTION
**Pull request by jameslamb**
Contributes to #4543, as part of #4310.

This PR proposes the following changes in the R package:

* deprecates `Dataset$getinfo()` (with a warning its name will be changed to `get_field()`)
* deprecates `Dataset$setinfo()` (with a warning its name will be changed to `set_field()`)
* adds method `Dataset$get_field(field_name)` to `Dataset`, matching the Python package
    - https://github.com/microsoft/LightGBM/blob/8a90ea3f267a81a529e3f069cc13e0f6320e7989/python-package/lightgbm/basic.py#L1939
* adds method `Dataset$set_field(field_name, data)` should be added to `Dataset`, matching the Python package
    - https://github.com/microsoft/LightGBM/blob/8a90ea3f267a81a529e3f069cc13e0f6320e7989/python-package/lightgbm/basic.py#L1890

### Notes for Reviewers

I only *added* new tests, but left existing tests with `getinfo()` and `setinfo()` in place, so that those methods are still tested as long as they exist in the package.

### CODE REVIEW

File: R-package/R/lgb.Dataset.R [hunk 1]
```diff
@@ -1173,6 +1205,8 @@ getinfo <- function(dataset, ...) {
 #' @export
 getinfo.lgb.Dataset <- function(dataset, name, ...) {
 
+  warning("Calling getinfo() on a lgb.Dataset is deprecated. Use set_field() instead.")
```
Comment (StrikerRUS) [inline R-package/R/lgb.Dataset.R:1208]:
```suggestion
  warning("Calling getinfo() on a lgb.Dataset is deprecated. Use get_field() instead.")
```

**Review by StrikerRUS — CHANGES_REQUESTED** [review]
Not sure whether we should update here
https://github.com/microsoft/LightGBM/blob/master/docs/FAQ.rst#i-used-setinfo-tried-to-print-my-lgb-dataset-and-now-the-r-console-froze. WDYT?

Update `Dataset$setinfo()` here 
https://github.com/microsoft/LightGBM/blob/54facc4d727812075b0c90e5506e521938953b25/R-package/R/lgb.Dataset.R#L557

File: R-package/R/lgb.Dataset.R [hunk 2]
```diff
@@ -1158,10 +1190,10 @@ slice.lgb.Dataset <- function(dataset, idxset, ...) {
 #' dtrain <- lgb.Dataset(train$data, label = train$label)
 #' lgb.Dataset.construct(dtrain)
 #'
-#' labels <- lightgbm::getinfo(dtrain, "label")
-#' lightgbm::setinfo(dtrain, "label", 1 - labels)
+#' labels <- lightgbm::get_field(dtrain, "label")
+#' lightgbm::set_field(dtrain, "label", 1 - labels)
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply R-package/R/lgb.Dataset.R:1194]: ah sure, you're right

**Comment by jameslamb** [comment]
> Update `Dataset$setinfo()` here

Done in https://github.com/microsoft/LightGBM/pull/4571/commits/a7459875a0fa42a827a0dbf82390ee346baa27b1

> Not sure whether we should update here
https://github.com/microsoft/LightGBM/blob/master/docs/FAQ.rst#i-used-setinfo-tried-to-print-my-lgb-dataset-and-now-the-r-console-froze. WDYT?

I think it's ok to leave that alone for now, since the functions are being deprecated but not removed yet. That entire FAQ section probably deserves some updates and revisions, to check if the issues listed there still exist in recent versions. But that could be done in a separate PR.

**Comment by jameslamb** [comment]
iiinteresting, I see this error on the `lint` job

> Some R documentation files have changed. Please re-generate them and commit those changes.

I did re-generate the docs before pushing. Maybe there was a new release of `{roxygen2}` or something. I'll investigate this later today!

**Comment by jameslamb** [comment]
> Maybe there was a new release of {roxygen2} or something

Yes, there was a new release (7.1.2) a few days ago: https://cran.r-project.org/web/packages/roxygen2/index.html.

But it also looks like there are a few local changes I forgot to `git add` 😆 

Regenerated the docs with `{roxygen}` 7.1.2 and added the requested changes in https://github.com/microsoft/LightGBM/pull/4571/commits/e92a28ed0033f097632f9fc7564ac446142e8adc

**Comment by shiyu1994** [comment]
Shall we also update here in the FAQ documentation? Will `set_field` encounter the same problem?
setinfo

**Comment by shiyu1994** [comment]
Other modifications LGTM. They seem to be simple renaming. And thanks @StrikerRUS for detecting the minor mistakes.

**Comment by jameslamb** [comment]
> Shall we also update here in the FAQ documentation? Will `set_field` encounter the same problem?
> setinfo

ah! I answered this in https://github.com/microsoft/LightGBM/pull/4571#issuecomment-924516705 actually. I don't think we need to worry about it for this PR. That particular question was added in #976 (4 years ago), so it might not even be relevant any more. I'd prefer to take updating the FAQ as a separate task.

**Review by StrikerRUS — APPROVED** [review]
LGTM! Thanks a lot for working on this!

**Pull request merged**